book/src/api/volume-snapshot.md

```diff
@@ -0,0 +1,1036 @@
+<em>
+string
+</em>
+</td>
+<td>
+<em>(Optional)</em>
+<p>name of the VolumeSnapshotClass from which this snapshot was (or will be)
+created.
+Note that after provisioning, the VolumeSnapshotClass may be deleted or
+recreated with different set of values, and as such, should not be referenced
+post-snapshot creation.</p>
+</td>
+</tr>
+<tr>
+<td>
+<code>source</code><br/>
+<em>
+<a href="#snapshot.storage.k8s.io/v1.VolumeSnapshotContentSource">
+VolumeSnapshotContentSource
+</a>
+</em>
+</td>
+<td>
+<p>source specifies whether the snapshot is (or should be) dynamically provisioned
+or already exists, and just requires a Kubernetes object representation.
+This field is immutable after creation.
+Required.</p>
+</td>
+</tr>
+<tr>
+<td>
+<code>sourceVolumeMode</code><br/>
+<em>
+<a href="https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#persistentvolumemode-v1-core">
+Kubernetes core/v1.PersistentVolumeMode
+</a>
+</em>
+</td>
+<td>
+<em>(Optional)</em>
+<p>SourceVolumeMode is the mode of the volume whose snapshot is taken.
+Can be either “Filesystem” or “Block”.
+If not specified, it indicates the source volume&rsquo;s mode is unknown.
+This field is immutable.
+This field is an alpha field.</p>
+</td>
+</tr>
+</table>
+</td>
+</tr>
+<tr>
+<td>
+<code>status</code><br/>
+<em>
+<a href="#snapshot.storage.k8s.io/v1.VolumeSnapshotContentStatus">
+VolumeSnapshotContentStatus
+</a>
+</em>
+</td>
+<td>
+<em>(Optional)</em>
+<p>status represents the current information of a snapshot.</p>
+</td>
+</tr>
+</tbody>
+</table>
+<h3 id="snapshot.storage.k8s.io/v1.DeletionPolicy">DeletionPolicy
+(<code>string</code> alias)</h3>
+<p>
+(<em>Appears on:</em><a href="#snapshot.storage.k8s.io/v1.VolumeSnapshotClass">VolumeSnapshotClass</a>, <a href="#snapshot.storage.k8s.io/v1.VolumeSnapshotContentSpec">VolumeSnapshotContentSpec</a>)
+</p>
+<div>
+<p>DeletionPolicy describes a policy for end-of-life maintenance of volume snapshot contents</p>
+</div>
+<table>
+<thead>
+<tr>
+<th>Value</th>
+<th>Description</th>
+</tr>
+</thead>
+<tbody><tr><td><p>&#34;Delete&#34;</p></td>
+<td><p>volumeSnapshotContentDelete means the snapshot will be deleted from the
+underlying storage system on release from its volume snapshot.</p>
+</td>
+</tr><tr><td><p>&#34;Retain&#34;</p></td>
+<td><p>volumeSnapshotContentRetain means the snapshot will be left in its current
+state on release from its volume snapshot.</p>
+</td>
+</tr></tbody>
+</table>
+<h3 id="snapshot.storage.k8s.io/v1.VolumeSnapshotContentSource">VolumeSnapshotContentSource
+</h3>
+<p>
+(<em>Appears on:</em><a href="#snapshot.storage.k8s.io/v1.VolumeSnapshotContentSpec">VolumeSnapshotContentSpec</a>)
+</p>
+<div>
+<p>VolumeSnapshotContentSource represents the CSI source of a snapshot.
+Exactly one of its members must be set.
+Members in VolumeSnapshotContentSource are immutable.
+TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotContentSource members
+will be immutable once specified.</p>
```

Comment:
Can you submit a PR in external-snapshotter to get rid of this TODO message?  It is no longer needed.
https://github.com/kubernetes-csi/external-snapshotter/blob/master/client/apis/volumesnapshot/v1/types.go#L344-L345

After that, re-run the script to generate the doc.

Reply:
I opened the PR at https://github.com/kubernetes-csi/external-snapshotter/pull/697.

Reply:
@xing-yang I have re-generated the doc to include the changes in https://github.com/kubernetes-csi/external-snapshotter/commit/b20011c84b1128ff21f432f3a448b63cc4b69f5b. PTAL.